psl-core/src/main/java/org/linqs/psl/reasoner/term/streaming/StreamingInitialRoundIterator.java

```diff
@@ -229,32 +235,47 @@ private GroundRule fetchNextGroundRule() {
             }
         }
 
+        // Template specific grounding
+        if (termCache.size() > 0) {
+            flushCache();
+        }
+
         currentRule++;
         if (currentRule >= rules.size()) {
             // There are no more rules, we are done.
+            currentRule--;
```

Comment:
Why the decrement here?
This path should directly initiate a `close()` on this iterator and shut down anymore access to this variable.